torch/export/_unlift.py

```diff
@@ -1,273 +1,322 @@
+                        if operand.target in state_dict:
+                            value = state_dict[operand.target]
+                        elif operand.target in constants:
+                            value = constants[operand.target]
                         else:
                             raise RuntimeError(
-                                f"Unable to find value for {operand.target}"
+                                "Unable to find value for ", operand.target
                             )
 
                         body_gm.register_buffer(operand.target, value)
                     else:
+                        submod_lifted_inputs.append(None)
                         real_operands.append(operand)
-                return real_operands
 
-            real_mapped_args = _find_real_operands(mapped_args, 0)
-            real_mapped_operands = _find_real_operands(operands, num_mapped)
+                return submod_lifted_inputs, real_operands
+
+            lifted_args, real_mapped_args = _unlift_submod_inputs(
+                body_gm, mapped_args, gm, unlifted_nodes
+            )
+            lifted_operands, real_mapped_operands = _unlift_submod_inputs(
+                body_gm, operands, gm, unlifted_nodes
+            )
+
             node.args = (body_graph, real_mapped_args, real_mapped_operands)
+            submod_lifted_inputs = lifted_args + lifted_operands
 
             _, in_spec = pytree.tree_flatten(real_mapped_args + real_mapped_operands)
 
+            # Currently HOO submodules do not allow mutations, so we
+            # will not need to handle this.
+            output_node = None
+            for node in gm.graph.nodes:
+                if node.op == "output":
+                    output_node = node
+                    break
+            assert output_node is not None
+            outputs = pytree.tree_flatten(output_node.args)[0]
+            mutated_outputs = [None for _ in range(len(outputs))]
+
             _unlift(
                 body_gm,
-                inp_pos_to_buffer_name_for_submod,
+                submod_lifted_inputs,
+                mutated_outputs,
                 in_spec,
                 None,
                 state_dict,
-                tensor_constants,
-                buffers_to_mutate,
+                constants,
             )
+
     gm.graph.lint()
     gm.graph.eliminate_dead_code()
     gm.recompile()
     return gm
 
 
-def _construct_inp_pos_to_param_buffer_name(
-    new_gm, graph_signature, state_dict, tensor_constants=None
-):
-    # TODO Fix the period in params/buffers names later
-    # maybe a pass to replace graph signature with fixed names
-    param_buffer_name_to_corrected_name = {}
-    constant_name_to_corrected_name = {}
-
+def _register_attrs_to_new_gm(
+    new_gm: torch.fx.GraphModule,
+    graph_signature: ExportGraphSignature,
+    state_dict: Dict[str, Any],
+    constants: Dict[str, Any],
+) -> None:
     for name, value in state_dict.items():
```

Comment:
one subtle thing that I noticed: for these, we are using the state dict as the source of truth for what is available, iterating over it and looking things up in the graph signature.

More properly it should be the other way around: we should treat the graph signature as the source of truth, and look up things in the state dict.

Why does this matter? Because the state dict comes from the user, so it could be wrong or have the wrong keys. The graph signature is our actual ground truth about what should go in the graph and not. 

So we should trust only the graph signature, and also double-check that the provided state dict matches. That includes catching the case where there is an extraneous key in the state dict, which in this code is passed through silently. 

This checking doesn't have to be here (doesn't have to be in this PR at all), maybe it should be earlier in the process. But something I noticed

Reply:
actually, I think I'll address this in a followup to tighten a few screws around this pass (including the buffer renaming stuff mentioned ovr chat). So you can ignore this feedback for this PR